### ISSUE TITLE
What to take in consideration when running samtools depth in parallel or concurrently?

### ISSUE THREAD

**Issue by tomkp75** [samtools/samtools#1242]
#### Is your feature request related to a problem? Please specify.
Hi,

I'm running several `samtools depth` commands on S3 and I'm hitting a bunch of `[E::bgzf_read] Read block operation failed with error 4`. If I run the same command independently on the same file it works fine.
What should I take in consideration in order to highly parallelize depth computation? I'm cutting into regions of a certain size (I've tried 1 million and up to 20 millions sized regions). Maybe I should consider cutting on a different basis like a number of blocks? How can I prevent samtools from crashing when computing multiple depth on the same file?

Thanks,

Tom

#### Describe the solution you would like.
A way to compute the depth per reference, in a highly distributed manner.

**Comment by tseemann** [comment]
What do you mean by running on S3 ?  
S3 is a block storage system - are you streaming BAMs from it?

**Comment by valeriuo** [comment]
Also, how do you spawn your parallel commands?

A simple Bash script that computes the depth for each region in a separate process and collates the results at the end went quite smoothly on both S3 and local FS. In fact, it got a 30% speed improvement on a 562k read sorted BAM file, comparing to a standard `samtools depth` command.

```c
f=/path/to/target-sorted.bam; declare -i i=0;
for r in $(samtools view $f | cut -f3 | uniq | sed "s/*//")
do 
    samtools depth $f -r $r > tmp.depth.$r.out &
    pids[$i]=$!
    regs[$i]=$r
    i=$i+1
done
for pid in ${pids[*]}; do wait $pid; done
for r in ${regs[*]}; do cat tmp.depth.$r.out >> depth.out; rm tmp.depth.$r.out; done
```

A footnote: when running parallel jobs on a S3 file, it is advisable to download the index file in the current folder a priori, otherwise each job might try to do that independently, resulting in a corrupt index file.

**Comment by tomkp75** [comment]
@tseemann Yes I mean running commands against objects stored on S3.
@valeriuo I spawned 32 processes, each one in charge of computing the depth of a different reference and each one running a _maximum_ of 12 asynchronous tasks for a maximum region task of 21,000,000 bps. That way 1 process could take on up to 250,000,000 and with 32 processes I could go over a bunch of references at once.
I ran my test against 3 BAM files of ~6G each and ~600 references total.
Dropping to 16 parallel processes and 12 concurrent tasks I got a decent result but I was using a bigger machine so I wanted to make use of all the processes I could.
As you mentioned, I had not downloaded the index file beforehand so it might have something to do with what you're suggesting 

> otherwise each job might try to do that independently, resulting in a corrupt index file.
